File: XMLCheck/Handler/CheckRulesXML01/Check_NGAY_VAO_NOI_TRU.cs
using System.Globalization;
using XMLCheck.Handler.CheckRulesXML1;
using XMLCheck.Models;
using XMLCheck.Services;

namespace XMLCheck.Handler.CheckRulesXML01
{
    public class Check_NGAY_VAO_NOI_TRU : Xml1RuleBase
    {
        public override string Key => "NGAY_VAO_NOI_TRU";

        public override ErrorDetails? Check(Xml1Model model, HoSoContext context)
        {
            var raw = model.NGAY_VAO_NOI_TRU_RAW;

            if (string.IsNullOrWhiteSpace(raw))
            {
                // Không bắt buộc → bỏ qua
                return null;
            }

            // 1. Quá kích thước tối đa (12)
            if (raw.Length > 12)
            {
                return Error("NGAY_VAO_NOI_TRU quá kích thước tối đa (12)");
            }

            // 2. Không đúng định dạng yyyyMMddHHmm
            if (!DateTime.TryParseExact(
                    raw,
                    "yyyyMMddHHmm",
                    null,
                    System.Globalization.DateTimeStyles.None,
                    out var ngayVaoNoiTru))
            {
                return Error("NGAY_VAO_NOI_TRU không đúng định dạng yyyyMMddHHmm");
            }

            // 3. Không được lớn hơn ngày hiện tại
            if (ngayVaoNoiTru > DateTime.Now)
            {
                return Error("NGAY_VAO_NOI_TRU không được lớn hơn ngày hiện tại");
            }

            return null;
        }
    }
}

File: XMLValidator_v2/src/parsers/rules/xml1/Check_NGAY_VAO_NOI_TRU.ts
import { Xml1RuleBase } from './xml1RuleBase';
import { Xml1Model } from '../../../interfaces/xml-models.interface';
import { HoSoContext } from '../../../services/hoso-context';
import { ErrorDetails } from '../../../dto/xml-check.dto';
import { parseExactDate } from '../../../utils/date.helper';

export class Check_NGAY_VAO_NOI_TRU extends Xml1RuleBase {
  constructor() {
    super({
      ruleId: 'RULE_XML1_NGAY_VAO_NOI_TRU_01',
      xmlType: 'XML1',
      field: 'NGAY_VAO_NOI_TRU',
      severity: 'ERROR',
      description: 'Kiểm tra quy định cho trường NGAY_VAO_NOI_TRU',
      errorMessage: 'NGAY_VAO_NOI_TRU quá kích thước tối đa (12)',
      reference: 'QĐ 3176/QĐ-BYT',
    });
  }

  

  public check(model: Xml1Model, context: HoSoContext): ErrorDetails | null {
    const raw = model.NGAY_VAO_NOI_TRU_RAW;

    if (!raw || !raw.trim()) {
      return null;
    }

    if (raw.length > 12) {
      return this.error('NGAY_VAO_NOI_TRU quá kích thước tối đa (12)');
    }

    const ngayVaoNoiTru = parseExactDate(raw);
    if (!ngayVaoNoiTru) {
      return this.error('NGAY_VAO_NOI_TRU không đúng định dạng yyyyMMddHHmm');
    }

    if (ngayVaoNoiTru > new Date()) {
      return this.error('NGAY_VAO_NOI_TRU không được lớn hơn ngày hiện tại');
    }

    return null;
  }
}
